versioned_docs/version-v4.0.0/developers/heartbeat.md

```
@@ -0,0 +1,35 @@
+
+A node's health score is increased by 0.1 when it responds to a ping with a pong packet or sends its own ping. Conversely, a node's health score is reduced by 0.1 if it does not respond to a ping. The health score ranges from 0 (minimum) to 1 (maximum). Nodes with a health score greater than or equal to the default health threshold (0.5) are considered online and suitable for use in paths. Nodes with lower health scores are omitted.
+
+### Exponential Backoff
+
+To account for the dynamic nature of network status changes (e.g., due to power outages or unstable network links), the HOPR network frequently remeasures node availability. However, constantly probing known offline nodes is inefficient. HOPR addresses this issue by using a heartbeat mechanism with exponential backoff.
+
+The time until the next ping (tbo) is sent to a node increases with the number of failed ping attempts (nfail) since the last successful attempt or the network start. A successful response to a ping resets the backoff timer.
+
+The backoff time (tbo) is calculated using the following formula:
+
+```
```

markdownlint-cli2 (0.22.1)

[warning] Fenced code blocks should have a language specified

(MD040, fenced-code-language)